.circleci/ci/ci.js

```diff
@@ -0,0 +1,167 @@
+    };
+
+    /**
+     * Switch working directory for the scope of the given function.
+     */
+    dir(dir, func) {
+        let currentDir = process.cwd();
+        process.chdir(dir);
+        console.log('// Changed directory to: ' + process.cwd());
+        try {
+            func();
+        } finally {
+            process.chdir(currentDir);
+            console.log('// Changed directory back to: ' + currentDir);
+        }
+    };
+
+    /**
+     * Checkout git repository with the given branch into the given folder.
+     */
+    checkout(repo, branch = 'master', folder = '') {
+        this.sh('git clone -b ' + branch + ' ' + repo + ' ' + folder);
+    };
+
+    /**
+     * Run shell command and attach to process stdio.
+     */
+    sh(command, returnStdout = false, print = true) {
+        if (print) {
+            console.log(command);
+        }
+        if (returnStdout) {
+            return e.execSync(command).toString().trim();
+        }
+        return e.execSync(command, {stdio: 'inherit'});
+    };
+
+    /**
+     * Return value of given environment variable.
+     */
+    env(key) {
+        return process.env[key];
+    }
+
+    /**
+     * Print stage name.
+     */
+    stage(name) {
+        console.log("\n------------------------------\n" +
+            "--\n" +
+            "-- %s\n" +
+            "--\n" +
+            "------------------------------\n", name);
+    };
+
+    /**
+     * Configure a git impersonation for the scope of the given function.
+     */
+    gitImpersonate(user, mail, func) {
+        try {
+            this.sh('git config --local user.name ' + user + ' && git config --local user.email ' + mail, false, false);
+            func()
+        } finally {
+            this.sh('git config --local --unset user.name && git config --local --unset user.email', false, false);
+        }
+    };
+
+    /**
+     * Configure git credentials for the scope of the given function.
+     */
+    gitCredentials(repo, func) {
+        try {
+            this.sh('git config credential.helper \'store --file .git-credentials\'');
+            fs.writeFileSync('.git-credentials', repo);
+            console.log('// Created file .git-credentials.');
+            func()
+        } finally {
+            this.sh('git config --unset credential.helper');
+            fs.unlinkSync('.git-credentials');
+            console.log('// Deleted file .git-credentials.');
+        }
+    };
+
+    /**
+     * Writes given content to a file.
+     */
+    writeFile(fileName, content) {
+        console.log(`// Write to file ${fileName}`);
+        fs.writeFileSync(fileName, content, { 'encoding': 'utf8' });
+    }
+
+    collectConfiguration() {
+        let configuration = {
+            modules: {}
+        };
+        
+        let folders = [process.cwd()];
+        process.stdout.write("Collecting project configuration");
+        while(folders.length) {
```

Comment:
It's a matter of taste, but `folders.forEach(folder => {` would more "standard", no?

Reply:
The loop adds subfolders to the `folders` array. `.forEach` does not work, when you manipulate the array while iterating through it.